designs/node-overlay.md

```diff
@@ -0,0 +1,305 @@
+
+* Custom resource requests: https://github.com/kubernetes-sigs/karpenter/issues/751
+* GPU time slicing: https://github.com/kubernetes-sigs/karpenter/issues/729
+* Resource Operators: https://github.com/kubernetes-sigs/karpenter/issues/1219
+* HugePage support: https://github.com/aws/karpenter-provider-aws/issues/6296
+
+While users have requested features to adjustment for node capacity on VM memory overhead, the initial release of Node Overlay will not include this functionality. The team is gathering additional data on memory overhead usage and other capacity-related use cases to inform the future API requirements. Future enhancements may include support for discovered existing capacity patterns, various node overhead configurations, and additional resource adjustment scenarios. These potential expansions, including memory overhead adjustments, will be evaluated based on users feedback and real-world usage patterns.
+
+* Support per provisioner(NodePool)/Instance type VM_MEMORY_OVERHEAD: https://github.com/aws/karpenter-provider-aws/issues/2141
+* Set System Reserved as percentage: https://github.com/kubernetes-sigs/karpenter/issues/1595
+
+## Proposal 
+
+NodeOverlays are a new Karpenter API that enable users to fine-tune the scheduling simulation for advanced use cases. NodeOverlays can be configured with a `requirements` which flexibly constrain when they are applied to a simulation. Requirements can match well-known labels like `[node.kubernetes.io/instance-type](http://node.kubernetes.io/instance-type)` or  `karpenter.sh/nodepool` and custom labels defined in a NodePool's `.spec.labels` or `.spec.requirements`. When multiple NodeOverlays match an offering for a given simulation, they are merged together using `spec.weight` to resolve conflicts.
+
+Previous attempts to implement instance type modifications focused on a 1:1 mapping approach, but this was abandoned after user feedback indicated that managing the combinatorial explosion of variations would be impractical for real-world use cases.
+
+```
+apiVersion: karpenter.sh/v1alpha1
+kind: NodeOverlay
+metadata:
+  name: default
+spec:
+  weight: ...
+  requirements:
+    ... 
+  priceAdjustment: ... 
+  capacity: 
+    ...
+status:
+  condition:
+  - lastTransitionTime: "..."
+    message: ""
+    reason: ...
+    status: "True"
+    type: Ready
+```
+
+### Example
+
+**Price Adjustment**: Define a pricing adjustment for all instance types option that fit the match labels field. User can set a scaler value or a percent
```

Comment:
Can you give more examples to make price-adjustment clearer. 

I'm guessing the current example means `90% of the OD price` (ie 10% off)? 

So a scaler would always be negative and look like `-0.1` or something?

Reply:
👍  you should explicitly state if it is an increase or decrease in price either with a sign (-/+) or a diff field for the sign

Reply:
Yeah, I added a section on how exactly this would reflect and removed the notion of floating point values